spec/index.bs

```diff
@@ -8668,50 +8669,59 @@ a list of vertices to process for each instance.
-
-        Note: in case of indirect draw calls, the `indexCount`, `vertexCount`,
-        and other properties of |drawCall| are read from the indirect buffer
-        instead of the draw command itself.
-
-        Issue: specify indirect commands better.
+    **Arguments:**
+    - |drawCall|: The draw call parameters.
+    - |state|: The active [=RenderState=].
+
+    **Returns:** list of integer indices.
+
+    1. Let |vertexIndexList| be an empty list of indices.
+    1. If |drawCall| is an indexed draw call:
+        1. Initialize the |vertexIndexList| with |drawCall|.indexCount integers.
+        1. For |i| in range 0 .. |drawCall|.indexCount (non-inclusive):
+            1. Let |relativeVertexIndex| be [$fetch index$](|i| + |drawCall|.`firstIndex`,
+                |state|.[=RenderState/indexBuffer=].`buffer`, |state|.[=RenderState/indexBuffer=].`offset`,
+                |state|.[=RenderState/indexBuffer=].`format`).
+            1. If |relativeVertexIndex| has the special value `"out of bounds"`,
+                stop and return the empty list.
+
+                Note: Implementations may choose to display a warning when this occurs,
+                especially when it is easy to detect (like in non-indirect indexed draw calls).
+            1. Append |drawCall|.`baseVertex` + |relativeVertexIndex| to the |vertexIndexList|.
+    1. Otherwise:
+        1. Initialize the |vertexIndexList| with |drawCall|.vertexCount integers.
+        1. Set each |vertexIndexList| item |i| to the value |drawCall|.firstVertex + |i|.
+    1. Return |vertexIndexList|.
+
+    Note: in case of indirect draw calls, the `indexCount`, `vertexCount`,
+    and other properties of |drawCall| are read from the indirect buffer
+    instead of the draw command itself.
+
+    Issue: specify indirect commands better.
 </div>
 
 <div algorithm>
     <dfn abstract-op>fetch index</dfn>(i, buffer, offset, format)
 
-        **Arguments:**
-            - |i|: Index of a vertex index to fetch.
-            - |buffer|: {{GPUBuffer}} containing index data.
-            - |offset|: Base offset into the |buffer|.
-            - |format|: {{GPUIndexFormat}} of the index.
+    **Arguments:**
+    - |i|: Index of a vertex index to fetch.
+    - |buffer|: {{GPUBuffer}} containing index data.
+    - |offset|: Base offset into the |buffer|.
+    - |format|: {{GPUIndexFormat}} of the index.
 
-        Let |stride| be defined by the |format|:
+    **Returns:** unsigned integer or `"out of bounds"`
+
+    1. Let |elementByteSize| be defined by the |format|:
         <dl class="switch">
             : {{GPUIndexFormat/"uint16"}}
             :: 2
             : {{GPUIndexFormat/"uint32"}}
             :: 4
         </dl>
-        Interpret the data in |buffer| starting with |offset| + |i| * |stride|
-        of size |stride| bytes as an unsigned integer and return it.
+    1. If |offset| + |i + 1| &times; |elementByteSize| &gt; |buffer|.{{GPUBuffer/[[size]]}},
```

Comment:
we shouldn't be comparing this against the buffer size. Instead, we should be considering the binding size for the index buffer, i.e. stuff passed in the last argument to `setIndexBuffer()`, if any